Homework/homework7/task2.py

```diff
@@ -0,0 +1,32 @@
+"""
+Given two strings.
+
+Return if they are equal when both are typed into
+empty text editors. # means a backspace character.
+Note that after backspacing an empty text, the text will continue empty.
+Examples:
+    Input: s = "ab#c", t = "ad#c"
+    Output: True
+    # Both s and t become "ac".
+    Input: s = "a##c", t = "#a#c"
+    Output: True
+    Explanation: Both s and t become "c".
+    Input: a = "a#c", t = "b"
+    Output: False
+    Explanation: s becomes "c" while t becomes "b".
+"""
+
+
+def backspaced_string(string: str) -> str:
+    new_string = ""
+    for char in string:
+        if char != "#":
+            new_string += char
+        else:
+            new_string = new_string[:-1]
+    return new_string
+
+
+def backspace_compare(first: str, second: str) -> bool:
```

Comment:
O( (2n-2) + (2k-2) ), where n and k - length of first and second strings (if I correctly understand). Because of slice.
So is it better to use a list as a `new_string` and than join it into the string, because a list has a `pop` method which is O(1)?

Reply:
```python
for char in string:  # O(N)
   new_string += char  # O(N)  we create a new object and copy all chars in it. 
```


The actual data you copy is half of the square:
```
a _ _
a b _
a b c
```

The complexity of this code is O(n^2/2)  which is written as O(n^2) and named quadratic.
Simple rule if something is happening inside the cycle you multiply.

Yes, list is a solution.

Reply:
Thank you fro explanation!